sound/soc/sof/ipc4-topology.c

```diff
@@ -1373,19 +1373,12 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		copier_data = &ipc4_copier->data;
 		available_fmt = &ipc4_copier->available_fmt;
 
-		/*
-		 * base_config->audio_fmt and out_audio_fmt represent the input and output audio
-		 * formats. Use the input format as the reference to match pcm params for playback
-		 * and the output format as reference for capture.
-		 */
-		if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
-			available_fmt->ref_audio_fmt = &available_fmt->base_config->audio_fmt;
-			ref_audio_fmt_size = sizeof(struct sof_ipc4_base_module_cfg);
-		} else {
-			available_fmt->ref_audio_fmt = available_fmt->out_audio_fmt;
-			ref_audio_fmt_size = sizeof(struct sof_ipc4_audio_format);
-		}
-
+		/*
+		 * base_config->audio_fmt represent the input audio formats. Use
+		 * the input format as the reference to match pcm params
+		 */
+		available_fmt->ref_audio_fmt = &available_fmt->base_config->audio_fmt;
+		ref_audio_fmt_size = sizeof(struct sof_ipc4_base_module_cfg);
```

Comment:
how do we know this is the right change? 

IIRC the code was tested precisely on capture by @bardliao, so why wasn't this caught earlier?

Lost in space and it's only Monday 9:30 am.

Reply:
is there an issue filed to show what was failing without this @RanderWang

Reply:
@plbossart  @ranj063  When I use buffer copier to convert format from S32_LE to S16_LE in DMIC capture, I found that the source module can't match up with this buffer copier since it use output format S16_LE as reference, not input. We didn't get it since we just use buffer copier as a simple copy, not format conversion.

Reply:
@plbossart @ranj063 I never changed audio format in the buffer type copier. i.e. The input and output format are always the same in my test. The issue @RanderWang met is that the output format of DMIC is S32_LE and he want to convert it to S16_LE in the buffer copier, and the existing code search S16_LE from available_fmt->out_audio_fmt and there is no S16_LE in the buffer copier's output available formats.
IMHO, the buffer type copier should always search audio format from input available format just like other modules like gain, mixer, etc. And `pipeline_params` should be changed to the output audio format. Note that `pipeline_params` is used to pass the audio format to the next widget, therefore it should always be the same as the previous widget's output format.

Reply:
Looks like people need a PhD in IPC4 to figure out how to deal with formats?

I for one have zero understanding of what these concepts are, nor what matching means or what the 'reference' is. can we get an explanation of what this is:

available_fmt->base_config->audio_fmt
available_fmt->out_audio_fmt 
available_fmt->in_audio_fmt

Reply:
> IMHO, the buffer type copier should always search audio format from input available format just like other modules like gain, mixer, etc. And `pipeline_params` should be changed to the output audio format. Note that `pipeline_params` is used to pass the audio format to the next widget, therefore it should always be the same as the previous widget's output format.

@bardliao @RanderWang this patch only changes the criterion for matching. Where aer we doing the part where we update the pipeline_params to reflect what the output format of the buffer copier is?

Reply:
@ranj063 yes, it is fixed by https://github.com/thesofproject/linux/pull/4097 ?  I didn't get this issue since the sink module is in another pipeline.

Reply:
@ranj063 I added some code to deal with output format but finally found we have processed this case in this function.
````c
static int
sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
			       struct snd_pcm_hw_params *fe_params,
			       struct snd_sof_platform_stream_params *platform_params,
			       struct snd_pcm_hw_params *pipeline_params, int dir)
{
...
	/* modify the input params for the next widget */
	fmt = hw_param_mask(pipeline_params, SNDRV_PCM_HW_PARAM_FORMAT);
	out_sample_valid_bits =
		SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(copier_data->out_format.fmt_cfg);
	snd_mask_none(fmt);
	switch (out_sample_valid_bits) {
	case 16:
		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S16_LE);
		break;
	case 24:
		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S24_LE);
		break;
	case 32:
		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
		break;
	default:
		dev_err(sdev->dev, "invalid sample frame format %d\n",
			params_format(pipeline_params));
		return -EINVAL;
	}
...
}
````

Reply:
I designed a pipeline with two buffer copiers to do format conversion from 32 to 16 then 16 to 32, everything works with this PR.
![sof-hda-generic](https://user-images.githubusercontent.com/39945651/212028961-90b3bedb-7770-4980-9154-f84d6568240b.png)